### ISSUE TITLE
Serve different responses based on user request payload

### ISSUE THREAD

**Issue by chrisahardie** [stoplightio/prism#1212]
**User story.**
As a user, I can select a localization language by subdomain, so that I can get proper translations for the subdomain I am visiting in my web app.

**Is your feature request related to a problem?**

My api serves localization files by subdomain. As I nav through different subdomains, a request is made like `/localizations?subdomain=account&language=en`

It would be convenient during development if I could mock out that endpoint but still serve particular localization payloads based on the `subdomain` and `language` querystring variables. That way my app would be properly localized.

**Describe the solution you'd like**

Some mechanism where I could supply multiple criteria to an endpoint to receive particular responses. I can presently add a `Prefer` header, but that appears to only allow me to select a response based on a single key e.g. `"Prefer: example=exampleKey"`

~Also it is mentioned that querystrings will be deprecated in favour of the `Prefer` header. Why is that? Presently, I supply values to my api via the querystring, there is no notion of the `Prefer` header in my API project. I would have to supply values differently depending on whether the app is in development or production mode.~ Resolved (@philsturgeon)

Thanks!

**Comment by XVincentX** [comment]
Hey @chrisahardie,

I do not really understand what you're asking for.

>  but that appears to only allow me to select a response based on a single key

What other criteria would you need to select your proper response?

> Why is that

Query strings interfere with the URL space that is hard to predict and should never be touched by Prism. It's not something we own. URL space is User space.

The same is the Prefer header, but it's a natural multi key/value parameters and isn't that common, so the collision risk is really minimal.

**Comment by chrisahardie** [comment]
Hi @XVincentX ,

Sorry I wasn't clear. My use case for prism is I need a mock Api to serve canned responses so I can develop a web app while our backend devs are still working on the Api. We can settle on the OpenApi specification for the api, so prism seems like it would be ideal for this purpose.

The mock api server needs to behave as the real Api will eventually behave, so I can properly code against it. As I navigate into various subdomains of my site - let's say the Account subdomain, the Cart subdomain, and the Contact Us subdomain - I request localizations (they are quite large, so we've chunked them up). For my localization endpoint, I need to supply 2 parameters like in the following examples:

```
/localization?subdomain=account&language=en
/localization?subdomain=cart&language=en
/localization?subdomain=contact-us&language=en
```

We also support French, so we would serve the French localizations with:

```
/localization?subdomain=account&language=fr
/localization?subdomain=cart&language=fr
/localization?subdomain=contact-us&language=fr
```

This means for my mock localization endpoint, I need to serve 1 of 6 canned responses. I need to identify the correct canned response by evaluating two criteria - subdomain and language. I don't see a way to do that presently.

If I understand properly, moving away from querystring variables is to prevent collisions? In other words, preclude collisions if the Api author has used `dynamic` or `example`? That seems like a valid decision, but in my case it would still be useful to interrogate querystring variables so as to select the proper canned response I want my endpoint to serve. Otherwise during development I'm going to add a header which I won't in production, so my mock Api server and production Api server aren't behaving the same and I'll need to code different flows for each.

```
sendRequest(opts) {
  if (process.env.NODE_ENV === "production") {
    return sendProdRequestWithQuerystring(opts);
  }

  return sendDevRequestWithPreferHeader(opts); // The Prefer header could vary with each request
}
```

If the behaviour I desire isn't presently available and isn't something you would consider for the core product, would it be behaviour that I would be able to effect once middleware is available?

Hope I was clearer! Thanks!

**Comment by XVincentX** [comment]
Oh ok I think I understand your use case now.

So the Prefer header is used to configure Prism preferences, such as example, dynamic and stuff like that. It does not decide anything based on your payload. This is essentially the reason why we decided to move that to the Header instead of the queystring. Querystring are way easier to get collided with other parameters — the Prefer header, well, just a little bit less.

To answer your question, it seems like you're interested in automatic responses based on the user payload. This is something that people has been asking but we never sat down and tried to figure out an implementation. We might in the future.

I can tell you for sure that:

1. This is not in our immediate short term roadmap
2. We are not planning in offering any middleware/extension capability at the moment

If this feature is really critical for you, I encourage you to fork the repository and make the necessary changes.

**Comment by chrisahardie** [comment]
Thanks @XVincentX , I appreciate you taking the time.

Shall I close this issue or leave it open as a feature request?

**Comment by XVincentX** [comment]
We can leave this issue open and see if we can do something; however nothing's gonna happen in the short period

**Comment by philsturgeon** [comment]
To summarize the basic request here is pretty much the same idea as #435 where people could use some sort of logic, code or config, to make /foo/X or /foo?x=y to mean that a specific response example should be chosen.

There's the idea of example pairs which we've discussed internally that only supports looking at URL parameters so that /foo/123 would pick an example called "/foo/123" from within the corresponding content type but that is an awkward convention to force on prism users who we also expect to be using Stoplight Docs as it will render those paths as a human readable name and... well that's not great.

If we do use this convention we would have to consider adding Summary and Description support for examples (https://github.com/stoplightio/studio/issues/508) so that we can hide this convention from users who happen to use Studio, making nice looking docs regardless of what docs tool people use (Swagger, ReDoc, Elements) yet still keep example pairs for Prism users.

@marbemac we've discussed this on and off for aaaages. Maybe we could talk about this at the next product meeting. :D
